activejob/lib/active_job/continuation.rb

```diff
@@ -0,0 +1,321 @@
+      else
+        run_step(name, start: start, &block)
+      end
+    end
+
+    def to_h
+      {
+        "completed" => completed.map(&:to_s),
+        "current" => current&.to_a
+      }.compact
+    end
+
+    def description
+      if current
+        current.description
+      elsif completed.any?
+        "after '#{completed.last}'"
+      else
+        "not started"
+      end
+    end
+
+    private
+      attr_reader :job, :encountered_step_names, :completed, :current
+
+      def advanced?
+        @advanced
+      end
+
+      def running_step?
+        @running_step
+      end
+
+      def started?
+        completed.any? || current.present?
+      end
+
+      def completed?(name)
+        completed.include?(name)
+      end
+
+      def validate_step!(name)
+        raise InvalidStepError, "Step '#{name}' must be a Symbol, found '#{name.class}'" unless name.is_a?(Symbol)
+        raise InvalidStepError, "Step '#{name}' has already been encountered" if encountered_step_names.include?(name)
+        raise InvalidStepError, "Step '#{name}' is nested inside step '#{current.name}'" if running_step?
+        raise InvalidStepError, "Step '#{name}' found, expected to resume from '#{current.name}'" if current && current.name != name && !completed?(name)
+
+        encountered_step_names << name
+      end
+
+      def new_step(*args, **options)
+        Step.new(*args, **options) { checkpoint! }
+      end
+
+      def skip_step(name)
+        instrument :step_skipped, step: name
+      end
+
+      def run_step(name, start:, &block)
+        @running_step = true
+        @current ||= new_step(name, start, resumed: false)
+
+        instrumenting_step(current) do
+          block.call(current)
+        end
+
+        @completed << current.name
+        @current = nil
+        @advanced = true
+
+        checkpoint!
+      ensure
+        @running_step = false
+        @advanced ||= current&.advanced?
+      end
+
+      def interrupt!
+        instrument_job :interrupt
+        raise Interrupt, "Interrupted #{description}"
+      end
+
+      def checkpoint!
+        interrupt! if job.queue_adapter.stopping?
+      end
+
+      def wrapping_errors_after_advancing(&block)
+        block.call
+      rescue StandardError => e
+        if !e.is_a?(Error) && advanced?
+          raise AfterAdvancingError, "Advanced job failed with error: #{e.message}"
+        else
+          raise
+        end
+      end
+
+      def instrumenting_step(step, &block)
+        instrument (step.resumed? ? :step_resumed : :step_started), step: step
+
+        block.call
+
+        instrument :step_completed, step: step
+      rescue Interrupt
+        instrument :step_interrupted, step: step
+        raise
+      end
+
+      def instrument_job(event)
+        instrument event, description: description, completed_steps: completed, current_step: current
+      end
+
+      def instrument(event, payload = {})
+        job.send(:instrument, event, **payload)
```

Comment:
It would be best to avoid `send` here, as it not good for JIT and such. The method can be public but marked as `# :nodoc:`.